### PR TITLE
Transaction fee refactor

### PR DESCRIPTION
**Pull request by QinshiWang**

### CODE REVIEW

File: contracts/WooFeeManager.sol [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+contract WooFeeManager is InitializableOwnable, ReentrancyGuard, IWooFeeManager {
+    /* ----- Type declarations ----- */
+
+    using SafeMath for uint256;
+    using DecimalMath for uint256;
+    using SafeERC20 for IERC20;
+
+    /* ----- State variables ----- */
+
+    mapping (address => uint256) public override feeRate;
```
Comment (skyinthesea-woo) [inline contracts/WooFeeManager.sol]: why is there a "override" ?

File: contracts/WooPP.sol [hunk 2]
```diff
@@ -69,6 +70,7 @@ contract WooPP is InitializableOwnable, ReentrancyGuard, Pausable, IWooPP {
     address public immutable override quoteToken;
     address public wooracle;
     IWooGuardian public wooGuardian;
+    address public feeManager;
```
Comment (skyinthesea-woo) [inline contracts/WooPP.sol:73]: Also need to init this in ctor, right ?

File: contracts/interfaces/IWooFeeManager.sol [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+* AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+* LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+* OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+*/
+
+/// @title Contract to collect transaction fee of Woo private pool.
+interface IWooFeeManager {
+    /* ----- Events ----- */
+
+    event FeeRateUpdated(address indexed token, uint256 newFeeRate);
```
Comment (skyinthesea-woo) [inline contracts/interfaces/IWooFeeManager.sol:42]: WooFeeRateUpdated

Comment (skyinthesea-woo) [reply contracts/interfaces/IWooFeeManager.sol:42]: Better to add "Woo" as prefix.

File: contracts/interfaces/IWooFeeManager.sol [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+    event Withdraw(address indexed token, address indexed to, uint256 amount);
```
Comment (skyinthesea-woo) [inline contracts/interfaces/IWooFeeManager.sol:43]: WooFeeWithdrawal

maybe this name ?

**Review by skyinthesea-woo — APPROVED** [review]

File: test/WooPP_part1.test.ts [hunk 5]
```diff
@@ -78,6 +80,7 @@ describe('WooPP Test Suite 1', () => {
     quoteToken = await deployMockContract(owner, IERC20.abi)
     baseToken1 = await deployMockContract(owner, IERC20.abi)
     baseToken2 = await deployMockContract(owner, IERC20.abi)
+    feeManager = await deployMockContract(owner, IWooFeeManager.abi)
```
Comment (skyinthesea-woo) [inline test/WooPP_part1.test.ts:83]: why we don't need to mock the methods of feeManager?

**Review by skyinthesea-woo — APPROVED** [review]

**Comment by skyinthesea-woo** [comment]
A few inline comments.

Others are good to go !

File: contracts/interfaces/IWooFeeManager.sol (hunk 3 above)
Comment (QinshiWang) [reply contracts/interfaces/IWooFeeManager.sol:42]: I don't think we should use `Woo` prefix for events. Events are bound to contracts. It is enough to only have `Woo` prefix for contracts.

File: contracts/interfaces/IWooFeeManager.sol (hunk 4 above)
Comment (QinshiWang) [reply contracts/interfaces/IWooFeeManager.sol:43]: This is copied from `WooPP`.

File: test/WooPP_part1.test.ts (hunk 5 above)
Comment (QinshiWang) [reply test/WooPP_part1.test.ts:83]: Without trading, we don't need to mock feeManager.

**Pull request merged**